Fix remote branch

rbailey-godaddy opened the pull request:
To help us get this pull request reviewed and merged quickly, please be sure to include the following items:

* [ ] Tests (if applicable)
* [ ] Documentation (if applicable)
* [X] Changelog entry
* [X] A full explanation here in the PR description of the work done

## PR Type
What kind of change does this PR introduce?

* [X] Bugfix
* [ ] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Build related changes
* [ ] CI related changes
* [ ] Documentation content changes
* [ ] Tests
* [ ] Other

## Backward Compatibility

Is this change backward compatible with the most recently released version? Does it introduce changes which might change the user experience in any way? Does it alter the API in any way?

* [X] Yes (backward compatible)
* [ ] No (breaking changes)


## Issue Linking
fixes #247 

## What's new?
- Silently prefix user-supplied branch name with remote source name (i.e. "origin/") when user specifies `--branch` option with `scan-remote-repo`. The source name typically will be "origin", but it's looked up when the repository is cloned and we'll use whatever ends up being defined.

rscottbailey committed:
Fix scan-remote-repo --branch

The underlying infrastructure requires the branch to be prefixed with
the remote (i.e. "origin/"). Because we know the clone will define the
target repository as "origin", we simply modify the branch reference
when setting up parameters and everything is good.

BTW, the last time this worked correctly was v2.0.2. Tartufo versions
v2.2.0 through v2.7.0 returned no errors, but also never found any
secrets, even when they should have done so. Beginning with v2.8.0,
you get an error stating the branch cannot be found.

Fixes: #247

rscottbailey committed:
Add changelog entry

rscottbailey committed:
Don't break non-branch operation

Only apply magic branch name fixup if there is a name to fix up;
otherwise, pass `None` through as-is.

rbailey-godaddy commented:
FWIW, I did a quick check (manually specifying `--branch origin/<branch>`) against v2.9 (`main`) -- this fix doesn't work, so if we wanted to backport we'd need to do some more clever (or at least with more involved research).

sushantmimani reviewed (approved):
LGTM 🦅

tarkatronic commented on tartufo/commands/scan_remote_repo.py:
```diff
@@ -53,10 +53,11 @@ def main(
     include_submodules: bool,
 ) -> GitRepoScanner:
     """Automatically clone and scan a remote git repository."""
+    target_branch = f"origin/{branch}" if branch else None
```
This might potentially re-introduce #203. Since we control the clone, can we be sure to specify the remote name as `origin` somehow? Or instead of hard-coding `origin` here, can we inspect the repository extract the remote name?

rbailey-godaddy replied:
Our clone process always names the remote source "origin" (I checked). If the underlying library changes, that could be a problem. I'll look at putting in a step to retrieve the remote name just in case.

tarkatronic replied:
My concern is, `GitPython` uses actual `git` commands under the surface, which ends up reading user config... so could a user have their config set up such that it automatically clones with a different name?

rbailey-godaddy replied:
I just added a commit that returns the actual remote source when the repository is cloned, and then uses that.

rscottbailey committed:
Don't guess remote repository name

Thus far, when `scan-remote-repo` clones the target into a scratch
directory, the target's remote reference will be `origin`. For purposes
of future-proofing, look up and return the origin name so the caller
does not have to assume it's "origin".

rscottbailey committed:
Missed adding to last commit :-p

rscottbailey committed:
Unit test tweaks

Try to make path generic (like following test) to avoid spurious
failures because we can't agree on path syntax.
